### ISSUE TITLE
`this` implies global scope in code, but not at run-time

### ISSUE THREAD

**Issue by ntilwalli** [rollup/rollup#1007]
More of a question than an issue... I'm trying to bundle `superagent` with rollup which is written in commonjs. It has a module which appends methods directly to the `exports` object which all use `this` directly. Those methods are then mixed-in (copied over) to a different object in other modules. 

In the original module the `this` looks like it's referencing the global scope so rollup sets `this` to `undefined` and at run-time calls to those methods error out because this has been explicitly reset
When I modify the code to assign the methods to a tmp object and assign that to the `exports` and then copy those methods from the that tmp object, since it no longer seems to be referencing the global scope `this` doesn't get assigned to `undefined` and all works fine

The offending module is here: https://github.com/visionmedia/superagent/blob/master/lib/request-base.js

The methods are copied over here: https://github.com/visionmedia/superagent/blob/master/lib/client.js#L524

What's the best approach to deal with this?

**Comment by ntilwalli** [comment]
Is the only way to PR the `superagent` repo so that module has more ES6 conformant module semantics?

**Comment by ntilwalli** [comment]
Would it be possible to add `allowGlobalThisReferences` configuration flag which accepts files where `this` is not set to undefined?

**Comment by TrySound** [comment]
Did you try `context` option?

**Comment by ntilwalli** [comment]
@TrySound Does `context` allow me to set the option for one file only?  I want `this` to be left alone so the functions can be later set on the `prototype` of other functions during run time, where they should have standard `this` semantics...  What would I set `context` to?

**Comment by Rich-Harris** [comment]
Closing via #1023

**Comment by ahelmberger** [comment]
I'm sorry, I tried to get my head around this for two hours, but still didn't get it to work with superagent. My config looks like this:

~~~
const { rollup } = require('rollup');
const nodeResolve = require('rollup-plugin-node-resolve');
const commonjs = require('rollup-plugin-commonjs');
const babel = require('rollup-plugin-babel');

module.exports = {
  entry: 'app/main.js',
  plugins: [
    nodeResolve({ jsnext: true, main: true, browser: true, preferBuiltins: false }),
    commonjs(),
    babel()
  ],
  moduleContext: {
    './node_modules/superagent/lib/request-base.js': 'window'
  },
  targets: [{
    dest: 'dist/main.js',
    format: 'iife',
    sourceMap: true,
    useStrict: true
  }]
};
~~~

`clearTimeout` seems to be not the global `clearTimeout` and I get the error. Any hint into the right direction would be appreciated. Thanks.

**Comment by ntilwalli** [comment]
@ahelmberger Yes, this continues to be an issue.  The merged PR actually did not fix it.  There are multiple issues I've posted about this..

https://github.com/rollup/rollup-plugin-commonjs/issues/127
https://github.com/rollup/rollup/issues/1027

For a quick and quite dirty workaround check out the changes associated with this unmerged superagent PR: https://github.com/visionmedia/superagent/pull/1076

**Comment by ahelmberger** [comment]
@ntilwalli Thanks for the clarification. That's quite an unsatisfying situation. I'll let you know if I can come up with another solution.

**Comment by ahelmberger** [comment]
@ntilwalli If you're interested, this solution is even quicker and dirtier than yours (doesn't require any forks), using (rollup/rollup-plugin-replace)[https://github.com/rollup/rollup-plugin-replace]:

~~~
const { rollup } = require('rollup');
const nodeResolve = require('rollup-plugin-node-resolve');
const commonjs = require('rollup-plugin-commonjs');
const replace = require('rollup-plugin-replace');
const babel = require('rollup-plugin-babel');

module.exports = {
  entry: 'app/main.js',
  plugins: [
    replace({
      include: '**/superagent/lib/request-base.js',
      values: { ' clearTimeout': ' window.clearTimeout' }
    }),
    nodeResolve({ jsnext: true, main: true, browser: true, preferBuiltins: false }),
    commonjs(),
    babel()
  ],
  targets: [{
    dest: 'dist/main.js',
    format: 'iife',
    sourceMap: true,
    useStrict: true
  }]
};
~~~

**Comment by ntilwalli** [comment]
@ahelmberger I like that! I never played with that plugin. Less invasive for sure.

**Comment by Rich-Harris** [comment]
This is fixed properly with rollup-plugin-commonjs 6.0.1
